url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/utils/Report.java

```diff
@@ -0,0 +1,386 @@
+		List<DBProtocol> protocols = new LinkedList<>();
+		BufferedReader reader = new BufferedReader(new FileReader(f));
+		String regex = "TMA-\\{(\\d+)\\}";
+		Pattern pattern = Pattern.compile(regex);
+		String retrievedText = null;
+		for (String line = reader.readLine(); line  != null; line = reader.readLine()){
+			Matcher m = pattern.matcher(line);
+			while (m.find()) {
+				retrievedText = m.group(1);
+				TMAProtocol p = new TMAProtocol();
+				p.setNumberOfRounds(Integer.parseInt(retrievedText));
+				protocols.add(p);
+			}
+		}
+		reader.close();
+		return protocols;
+	}
+	private static List<DBProtocol> getSwissKnifeProtocol(File f) throws NumberFormatException, IOException {
+		List<DBProtocol> protocols = new LinkedList<>();
+		BufferedReader reader = new BufferedReader(new FileReader(f));
+		String regex = "SwissKnife-\\{(\\d+)\\}";
+		Pattern pattern = Pattern.compile(regex);
+		String retrievedText = null;
+		for (String line = reader.readLine(); line  != null; line = reader.readLine()){
+			Matcher m = pattern.matcher(line);
+			while (m.find()) {
+				retrievedText = m.group(1);
+				SwissKnifeProtocol p = new SwissKnifeProtocol();
+				p.setNumberOfRounds(Integer.parseInt(retrievedText));
+				protocols.add(p);
+			}
+		}
+		reader.close();
+		return protocols;
+	}
+
+	private static List<DBProtocol> getSKIProtocol(File f) throws NumberFormatException, IOException {
+		List<DBProtocol> protocols = new LinkedList<>();
+		BufferedReader reader = new BufferedReader(new FileReader(f));
+		String regex = "SKI-\\{(\\d+), (\\d+)\\}";
+		Pattern pattern = Pattern.compile(regex);
+		String retrievedText = null;
+		for (String line = reader.readLine(); line  != null; line = reader.readLine()){
+			Matcher m = pattern.matcher(line);
+			while (m.find()) {
+				SKIProtocol p = new SKIProtocol(Integer.parseInt(m.group(2)), DBProtocol.SIZE_OF_NONCES);
+				p.setNumberOfRounds(Integer.parseInt(m.group(1)));
+				protocols.add(p);
+			}
+		}
+		reader.close();
+		return protocols;
+	}
+
+	
+	private static List<DBProtocol> getRasmussenAndCapckunProtocol(File f) throws NumberFormatException, IOException {
+		List<DBProtocol> protocols = new LinkedList<>();
+		BufferedReader reader = new BufferedReader(new FileReader(f));
+		String regex = "RC-\\{(\\d+)\\}";
+		Pattern pattern = Pattern.compile(regex);
+		String retrievedText = null;
+		for (String line = reader.readLine(); line  != null; line = reader.readLine()){
+			Matcher m = pattern.matcher(line);
+			while (m.find()) {
+				retrievedText = m.group(1);
+				RasmussenAndCapckunProtocol p = new RasmussenAndCapckunProtocol();
+				p.setNumberOfRounds(Integer.parseInt(retrievedText));
+				protocols.add(p);
+			}
+		}
+		reader.close();
+		return protocols;
+	}
+
+	private static List<DBProtocol> getPoulidorProtocol(File f) throws NumberFormatException, IOException {
+		List<DBProtocol> protocols = new LinkedList<>();
+		BufferedReader reader = new BufferedReader(new FileReader(f));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 239, 243. The resource is closed at line 252. There are other execution paths that do not contain closure statements, for example, when Integer.parseInt throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).